packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/BillingClientFactoryImpl.java

```diff
@@ -15,11 +18,34 @@ final class BillingClientFactoryImpl implements BillingClientFactory {
 
   @Override
   public BillingClient createBillingClient(
-      @NonNull Context context, @NonNull MethodChannel channel, int billingChoiceMode) {
+      @NonNull Context context,
+      @NonNull MethodChannel channel,
+      int billingChoiceMode,
+      @Nullable UserChoiceBillingListener userChoiceBillingListener) {
     BillingClient.Builder builder = BillingClient.newBuilder(context).enablePendingPurchases();
-    if (billingChoiceMode == BillingChoiceMode.ALTERNATIVE_BILLING_ONLY) {
-      // https://developer.android.com/google/play/billing/alternative/alternative-billing-without-user-choice-in-app
-      builder.enableAlternativeBillingOnly();
+    switch (billingChoiceMode) {
+      case BillingChoiceMode.ALTERNATIVE_BILLING_ONLY:
+        // https://developer.android.com/google/play/billing/alternative/alternative-billing-without-user-choice-in-app
+        builder.enableAlternativeBillingOnly();
+        break;
+      case BillingChoiceMode.USER_CHOICE_BILLING:
+        if (userChoiceBillingListener != null) {
+          // https://developer.android.com/google/play/billing/alternative/alternative-billing-with-user-choice-in-app
+          builder.enableUserChoiceBilling(userChoiceBillingListener);
+        } else {
+          Log.e(
+              "BillingClientFactoryImpl",
```

Comment:
What do you think about enforcing this being an error case, instead of defaulting? I think I might be confusing for developers to default to `PLAY_BILLING_ONLY`, even if they have made a mistake in how they configured for `USER_CHOICE_BILLING`. 

Not a strong preference, though, as long as the messaging is clear

Reply:
After reading further, we should never reach this case right? We define the listener, and we basically do it as a pass through where we pass the `UserChoiceDetails` in a copied hashmap to the dart side?

And the only time that listener is null is when the `BillingChoiceMode` is not `USER_CHOICE_BILLING`?

Can't we instead move that logic where we construct the listener (lines 547-557 in `MethodCallHandlerImpl`) inside this class and only invoke it when we actually need to make the listener, so it is never null?

Reply:
I agree that we should not every hit this case as the code is currently written but I added the additional fallaback logic to protect against future bugs and give us a way to know that we were in a bad state. 

I didnt want to create the listener in this class because it needed to invoke methodChannel and I thought the methodchannelimpl class would be where you would look for that logic.

Reply:
Part of why I suggested this is that we have existing code where we do something similar to construct a listener in this plugin, see https://github.com/flutter/packages/blob/a10b360a21531d03fb3795b7c083dda5c4796f96/packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/PluginPurchaseListener.java#L39.

I still lean towards moving the listener construction, so we can eliminate the bad case here, but if you have a strong preference to keep the structure as is I'm fine that way too

Reply:
How about a compromise. I am worried about the march 13th deadline for apps to 
 migrate to this api. I have filed https://github.com/flutter/flutter/issues/144851 and assigned it to myself along with some other cleanup bugs that I can do without impacting the public facing api.